docs/design-patterns.mdx

```diff
@@ -74,75 +180,279 @@ and in the transaction that is used to deploy it:
+by using `&AnyResource{Receiver}` instead of `&{Receiver}`.
+In the latter case, `AnyResource` is implicit, but not as clear as the former case.
+
+## Plural Names For Arrays And Maps Are Preferable
+
+e.g. `accounts` rather than `account` for an array of accounts.
+
+This signals that the field or variable is not scalar.
+It also makes it easier to use the singular form for a variable name during iteration.
+
+## Use transaction Post-Conditions when Applicable
 
 ### Problem
 
-Your contract, resource or struct has a resource that you wish to access fields of off-chain via a script. But scripts cannot return resources.
+Transactions can contain any amount of valid Cadence code and access many contracts and accounts.
+The power of resources and capabilities means that there may be some behaviors of programs that are not expected.
 
 ### Solution
 
-Declare a struct to hold the data that you wish to return from the script. Write a function that fills out the fields of this struct with the data from the resource that you wish to access. Then call this on the resource that you wish to access the fields of in a script, and return the struct from the script.
+It is usually safe to include post-conditions in transactions to verify the intended outcome.
 
-See [Script-Accessible Public Field](#script-accessible-public-field), above, for how best to expose this capability.
+### Example
 
-### Example Code
+This could be used when purchasing an NFT to verify that the NFT was deposited in your account's collection.
 
 ```cadence
-pub contract AContract {
-    pub struct BReportStruct {
-        pub var c: UInt64
-        pub var d: string
+// Psuedo-code
 
-        init(c: UInt64, d: string) {
-            self.c = c
-            self.d = d
-        }
+transaction {
 
+    pub let buyerCollectionRef: &NonFungibleToken.Collection
+
+    prepare(acct: AuthAccount) {
+
+        // Get tokens to buy and a collection to deposit the bought NFT to
+        let temporaryVault <- vaultRef.withdraw(amount: 10.0)
+        let self.buyerCollectionRef = acct.borrow(from: /storage/Collection)
+
+        // purchase, supplying the buyers collection reference
+        saleRef.purchase(tokenID: 1, recipient: self.buyerCollectionRef, buyTokens: <-temporaryVault)
+
+    }
+    post {
+        // verify that the buyer now owns the NFT
+        self.buyerCollectionRef.idExists(1) == true: "Bought NFT ID was not deposited into the buyers collection"
     }
+}
+```
 
-    pub resource BResource {
-        pub var c: UInt64
-        pub var d: string
+## Avoid Excessive Load and Save Storage Operations (Prefer in-place Mutations)
 
-        pub fun generateReport() BReportStruct {
-            return BReportStruct(c: c, d: d)
-        }
+### Problem
+
+When modifying data in account storage, `load()` and `save()` are costly operations.
+This can quickly cause your transaction to reach the gas limit or slow down the network.
+
+### Solution
```

Comment:
Maybe also describe how this is not limited to direct storage API usage, but also e.g. contract objects and their fields (which are implicitly stored in storage, i.e. read from/written to), or nested resourced. 

For example, a collection contains a dictionary of NFTs. There is no need to move the whole dictionary out of the field, update the dictionary on the stack (e.g. adding or removing an NFT), and then move the whole dictionary back to the field, it can be updated in-place. The same goes for a more complex data structure like the sharded collection: It can be updated in-place by taking a reference to the nested object.